torchtune/data/_utils.py

```diff
@@ -142,3 +150,181 @@ def format_content_with_images(
             final_content_list.append({"type": "image", "content": images.pop(0)})
 
     return final_content_list
+
+
+def chain(*funcs: List[Callable]) -> Callable:
+    """
+    Chain a list of functions together into a single function.
+
+    Args:
+        *funcs (List[Callable]): list of functions to chain together
+
+    Returns:
+        Callable: chained function
+    """
+
+    def chained_fn(x):
+        for fn in funcs:
+            x = fn(x)
+        return x
+
+    return chained_fn
+
+
+@requires_torchdata
+def load_hf_dataset(
+    source: str,
+    transform: Transform,
+    filter_fn: Optional[Callable] = None,
+    shuffle: bool = True,
+    seed: int = 0,
+    num_workers: int = 0,
+    parallel_method: Literal["process", "thread"] = "thread",
+    streaming: bool = False,
+    **load_dataset_kwargs: Dict[str, Any],
+) -> DatasetType:
+    """
+    Load a HuggingFace dataset (Map or Streaming) and apply a Transform to it.
+
+    Args:
+        source (str): HuggingFace dataset source.
+        transform (Transform): Transform to apply to the samples of the dataset.
+        filter_fn (Optional[Callable]): Filter function to pass to HuggingFace dataset.
+        shuffle (bool): Whether to shuffle the dataset. Default is True. For streaming datasets, this is passed to
+            HuggingFace dataset as .shuffle(). For map datasets, a DistributedSampler is used.
+        seed (int): Seed for the random number generator in the case of Map style dataset shuffling. Default is 0.
+        num_workers (int): Number of workers to use for loading the dataset. Default is 0 (no parallelism). Setting this
+            greater than 0 will create `parallel_method` workers to perform transforms to the dataset.
+        parallel_method (Literal["process", "thread"]): Method to use for parallelism. Default is "thread". No effect if
+            num_workers is 0.
+        streaming (bool): whether to load a streaming vs map-style dataset. Default False.
+        **load_dataset_kwargs (Dict[str, Any]): Additional Keyword arguments to pass to HuggingFace dataset. See Hugging Face's
+            documentation.
+
+    Returns:
+        A ``torchdata.nodes`` iterator that can be passed directly to a Loader, or combined with other-datasets in a multi-dataset
+        sampler.
+    """
+    from torchdata.nodes import IterableWrapper, ParallelMapper, SamplerWrapper
+
+    # TODO: Remove lazy import when we can
+    # see: https://github.com/pytorch/torchtune/issues/2151
+    from torchtune.training._distributed import get_world_size_and_rank
```

Comment:
Realized the comment I responded to was outdated, so just in case you miss it: can now use `from torchtune.utils import get_world_size_and_rank` at the top of the file